File: frontend/src/api/invocationEvidence.ts
import type { ConnectorBoundedInvocationInventoryItem } from "./boundedInvocations";
import { ApiRequestError, apiFetch } from "./client";

export type ConnectorInvocationEvidenceInventoryItem = {
  ingestion_id: string;
  schema_version: "atlas.connector-invocation-evidence-ingestion.v1";
  version: 1;
  source_invocation_id: string;
  source_invocation_digest: string;
  package_digest: string;
  capability_id: string;
  capability_class: "C0" | "C1";
  required_permission: string;
  normalized_redacted_result_digest: string;
  evidence_package_id: string;
  evidence_schema_version: string;
  evidence_content_digest: string;
  evidence_metadata_digest: string;
  classification: string;
  retention_policy_id: string;
  retention_policy_digest: string;
  ingestion_policy_id: string;
  ingestion_policy_digest: string;
  ingestion_policy_version: string;
  evidence_item_count: number;
  evidence_bytes: number;
  observed_from: string;
  observed_to: string;
  ingested_at: string;
  instance_state: "enabled_invocation_evidence_ingested";
  canonical_digest: string;
  source_invocation_completed: true;
  evidence_ingested: true;
  immutable_storage_confirmed: true;
  encrypted_at_rest: true;
  transient_buffers_erased: true;
  artifact_channel_closed: true;
  knowledge_item_created: false;
  retrieval_published: false;
  model_context_available: false;
  graph_updated: false;
  scheduled: false;
  workflow_continued: false;
  execution_authorized: false;
  deployment_approved: false;
  infrastructure_mutation_performed: false;
  reused: boolean;
};

export type ConnectorInvocationEvidenceOption = {
  source_invocation_id: string;
  source_invocation_digest: string;
  capability_id: string;
  capability_class: "C0" | "C1";
  required_permission: string;
  ingestion_policy_id: string;
  ingestion_policy_digest: string;
  ingestion_policy_version: string;
  ingestion_policy_expires_at: string;
  classification: string;
  retention_policy_id: string;
  required_assurance_level: "single_factor" | "multi_factor" | "hardware_backed";
  maximum_evidence_items: number;
  maximum_evidence_bytes: number;
  resulting_instance_state: "enabled_invocation_evidence_ingested";
  irreversible_claim_required: true;
  automatic_retry_allowed: false;
  knowledge_item_created: false;
  retrieval_published: false;
  model_context_available: false;
  graph_updated: false;
  scheduled: false;
  workflow_continued: false;
  execution_authorized: false;
  deployment_approved: false;
  infrastructure_mutation_performed: false;
};

// Retained for the later evidence-to-knowledge migration, which still consumes the internal record.
export type ConnectorInvocationEvidence = ConnectorInvocationEvidenceInventoryItem & {
  claim_id: string;
  organization_id: string;
  environment_id: string;
  connector_id: string;
  release_version: string;
  manifest_digest: string;
  instance_id: string;
  instance_key: string;
  display_name: string;
  output_schema_digest: string;
  result_policy_digest: string;
  access_policy_id: string;
  access_policy_digest: string;
  encryption_profile_id: string;
  encryption_profile_digest: string;
  ingestion_adapter_id: string;
  ingested_by: string;
  purpose: string;
};

const stableId = /^[a-z][a-z0-9_.:-]{2,127}$/;
const digest = /^[a-f0-9]{64}$/;
const inventoryFields = new Set([
  "ingestion_id", "schema_version", "version", "source_invocation_id",
  "source_invocation_digest", "package_digest", "capability_id", "capability_class",
  "required_permission", "normalized_redacted_result_digest", "evidence_package_id",
  "evidence_schema_version", "evidence_content_digest", "evidence_metadata_digest",
  "classification", "retention_policy_id", "retention_policy_digest", "ingestion_policy_id",
  "ingestion_policy_digest", "ingestion_policy_version", "evidence_item_count",
  "evidence_bytes", "observed_from", "observed_to", "ingested_at", "instance_state",
  "canonical_digest", "source_invocation_completed", "evidence_ingested",
  "immutable_storage_confirmed", "encrypted_at_rest", "transient_buffers_erased",
  "artifact_channel_closed", "knowledge_item_created", "retrieval_published",
  "model_context_available", "graph_updated", "scheduled", "workflow_continued",
  "execution_authorized", "deployment_approved", "infrastructure_mutation_performed", "reused",
]);
const optionFields = new Set([
  "source_invocation_id", "source_invocation_digest", "capability_id",
  "capability_class", "required_permission",
  "ingestion_policy_id", "ingestion_policy_digest", "ingestion_policy_version",
  "ingestion_policy_expires_at", "classification", "retention_policy_id",
  "required_assurance_level",
  "maximum_evidence_items", "maximum_evidence_bytes", "resulting_instance_state",
  "irreversible_claim_required", "automatic_retry_allowed", "knowledge_item_created",
  "retrieval_published", "model_context_available", "graph_updated", "scheduled",
  "workflow_continued", "execution_authorized", "deployment_approved",
  "infrastructure_mutation_performed",
]);

function hasExactFields(value: Record<string, unknown>, fields: ReadonlySet<string>): boolean {
  const keys = Object.keys(value);
  return keys.length === fields.size && keys.every((key) => fields.has(key));
}

function isTimestamp(value: unknown): value is string {
  return typeof value === "string" && value.length <= 64 && !Number.isNaN(Date.parse(value));
}

function envelopeData(payload: unknown): unknown {
  if (!payload || typeof payload !== "object") {
    throw new Error("Invocation evidence response envelope is unsafe");
  }
  const envelope = payload as Record<string, unknown>;
  if (!hasExactFields(envelope, new Set(["data", "meta"]))) {
    throw new Error("Invocation evidence response envelope is unsafe");
  }
  const meta = envelope.meta;
  if (!meta || typeof meta !== "object" ||
    !hasExactFields(meta as Record<string, unknown>, new Set(["correlation_id", "generated_at"]))) {
    throw new Error("Invocation evidence response envelope is unsafe");
  }
  const responseMeta = meta as Record<string, unknown>;
  if (typeof responseMeta.correlation_id !== "string" ||
    responseMeta.correlation_id.length < 1 || responseMeta.correlation_id.length > 128 ||
    !isTimestamp(responseMeta.generated_at)) {
    throw new Error("Invocation evidence response envelope is unsafe");
  }
  return envelope.data;
}

function hasPreservedBoundary(record: Record<string, unknown>): boolean {
  return record.source_invocation_completed === true && record.evidence_ingested === true &&
    record.immutable_storage_confirmed === true && record.encrypted_at_rest === true &&
    record.transient_buffers_erased === true && record.artifact_channel_closed === true &&
    record.knowledge_item_created === false && record.retrieval_published === false &&
    record.model_context_available === false && record.graph_updated === false &&
    record.scheduled === false && record.workflow_continued === false &&
    record.execution_authorized === false && record.deployment_approved === false &&
    record.infrastructure_mutation_performed === false;
}

function hasSafeOptionBoundary(record: Record<string, unknown>): boolean {
  return record.irreversible_claim_required === true && record.automatic_retry_allowed === false &&
    record.knowledge_item_created === false && record.retrieval_published === false &&
    record.model_context_available === false && record.graph_updated === false &&
    record.scheduled === false && record.workflow_continued === false &&
    record.execution_authorized === false && record.deployment_approved === false &&
    record.infrastructure_mutation_performed === false;
}

function isInventoryItem(value: unknown): value is ConnectorInvocationEvidenceInventoryItem {
  if (!value || typeof value !== "object") return false;
  const record = value as Record<string, unknown>;
  return hasExactFields(record, inventoryFields) && [
    "ingestion_id", "source_invocation_id", "capability_id", "required_permission",
    "evidence_package_id", "evidence_schema_version", "classification", "retention_policy_id",
    "ingestion_policy_id", "ingestion_policy_version",
  ].every((field) => typeof record[field] === "string" && stableId.test(record[field])) &&
    record.schema_version === "atlas.connector-invocation-evidence-ingestion.v1" &&
    record.version === 1 && (record.capability_class === "C0" || record.capability_class === "C1") &&
    [
      "source_invocation_digest", "package_digest", "normalized_redacted_result_digest",
      "evidence_content_digest", "evidence_metadata_digest", "retention_policy_digest",
      "ingestion_policy_digest", "canonical_digest",
    ].every((field) => typeof record[field] === "string" && digest.test(record[field])) &&
    Number.isInteger(record.evidence_item_count) && (record.evidence_item_count as number) >= 1 &&
    (record.evidence_item_count as number) <= 1000 && Number.isInteger(record.evidence_bytes) &&
    (record.evidence_bytes as number) >= 0 && (record.evidence_bytes as number) <= 1_048_576 &&
    isTimestamp(record.observed_from) && isTimestamp(record.observed_to) &&
    isTimestamp(record.ingested_at) && Date.parse(record.observed_to) >=
      Date.parse(record.observed_from) && Date.parse(record.ingested_at) >=
      Date.parse(record.observed_to) &&
    record.instance_state === "enabled_invocation_evidence_ingested" &&
    typeof record.reused === "boolean" && hasPreservedBoundary(record);
}

function isOption(value: unknown): value is ConnectorInvocationEvidenceOption {
  if (!value || typeof value !== "object") return false;
  const record = value as Record<string, unknown>;
  return hasExactFields(record, optionFields) && [
    "source_invocation_id", "capability_id", "required_permission", "ingestion_policy_id",
    "ingestion_policy_version", "classification", "retention_policy_id",
  ].every((field) => typeof record[field] === "string" && stableId.test(record[field])) && [
    "source_invocation_digest", "ingestion_policy_digest",
  ].every((field) => typeof record[field] === "string" && digest.test(record[field])) &&
    (record.capability_class === "C0" || record.capability_class === "C1") &&
    isTimestamp(record.ingestion_policy_expires_at) &&
    (record.required_assurance_level === "single_factor" ||
      record.required_assurance_level === "multi_factor" ||
      record.required_assurance_level === "hardware_backed") &&
    Number.isInteger(record.maximum_evidence_items) &&
    (record.maximum_evidence_items as number) >= 1 &&
    (record.maximum_evidence_items as number) <= 1000 &&
    Number.isInteger(record.maximum_evidence_bytes) &&
    (record.maximum_evidence_bytes as number) >= 1 &&
    (record.maximum_evidence_bytes as number) <= 1_048_576 &&
    record.resulting_instance_state === "enabled_invocation_evidence_ingested" &&
    hasSafeOptionBoundary(record);
}

export async function getConnectorInvocationEvidence(input: {
  sourceInvocationId: string;
}): Promise<ConnectorInvocationEvidenceInventoryItem[]> {
  const sourceId = input.sourceInvocationId;
  if (!stableId.test(sourceId)) throw new Error("Exact bounded-invocation scope is required");
  const parameters = new URLSearchParams({ source_invocation_id: sourceId });
  const response = await apiFetch(
    `/api/v1/connectors/invocation-evidence?${parameters.toString()}`,
    { headers: { Accept: "application/json" } },
  );
  if (!response.ok) {
    throw new ApiRequestError("Invocation evidence inventory failed", response.status);
  }
  const payload: unknown = await response.json();
  const data = envelopeData(payload);
  if (!Array.isArray(data)) throw new Error("Invocation evidence inventory returned unsafe records");
  const records: ConnectorInvocationEvidenceInventoryItem[] = [];
  for (const candidate of data as unknown[]) {
    if (!isInventoryItem(candidate)) {
      throw new Error("Invocation evidence inventory returned unsafe records");
    }
    if (candidate.source_invocation_id !== sourceId) {
      throw new Error("Invocation evidence inventory crossed the requested invocation scope");
    }
    records.push(candidate);
  }
  return records;
}

export async function getConnectorInvocationEvidenceOptions(
  sourceInvocationId: string,
): Promise<ConnectorInvocationEvidenceOption[]> {
  if (!stableId.test(sourceInvocationId)) {
    throw new Error("Exact bounded-invocation scope is required");
  }
  const parameters = new URLSearchParams({ source_invocation_id: sourceInvocationId });
  const response = await apiFetch(
    `/api/v1/connectors/invocation-evidence/options?${parameters.toString()}`,
    { headers: { Accept: "application/json" } },
  );
  if (!response.ok) {
    throw new ApiRequestError("Invocation evidence options failed", response.status);
  }
  const payload: unknown = await response.json();
  const data = envelopeData(payload);
  if (!Array.isArray(data)) throw new Error("Invocation evidence options returned unsafe evidence");
  const options: ConnectorInvocationEvidenceOption[] = [];
  for (const candidate of data as unknown[]) {
    if (!isOption(candidate) || candidate.source_invocation_id !== sourceInvocationId) {
      throw new Error("Invocation evidence options returned unsafe evidence");
    }
    options.push(candidate);
  }
  return options;
}

export async function createConnectorInvocationEvidence(input: {
  invocation: ConnectorBoundedInvocationInventoryItem;
  option: ConnectorInvocationEvidenceOption;
  purpose: string;
}) {
  const { invocation, option, purpose } = input;
  if (!invocation.authorization_consumed || !invocation.capability_invoked ||
    !invocation.result_validated || !invocation.result_redacted ||
    !invocation.target_session_closed || !invocation.lease_revocation_confirmed ||
    invocation.target_connected || invocation.evidence_ingested ||
    invocation.instance_state !== "enabled_bounded_capability_invocation_completed" ||
    option.source_invocation_id !== invocation.invocation_id ||
    option.source_invocation_digest !== invocation.canonical_digest ||
    option.capability_id !== invocation.capability_id ||
    option.capability_class !== invocation.capability_class ||
    option.required_permission !== invocation.required_permission ||
    purpose.trim().length < 20 || purpose.length > 1000) {
    throw new Error("Exact current invocation-evidence option is required");
  }
  const response = await apiFetch("/api/v1/connectors/invocation-evidence", {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      "Idempotency-Key": `connector-invocation-evidence.${crypto.randomUUID()}`,
    },
    body: JSON.stringify({
      schema_version: "atlas.connector-invocation-evidence-input.v1",
      source_invocation_id: option.source_invocation_id,
      source_invocation_digest: option.source_invocation_digest,
      ingestion_policy_id: option.ingestion_policy_id,
      ingestion_policy_digest: option.ingestion_policy_digest,
      purpose: purpose.trim(),
      acknowledged_ingestion_is_one_way_and_does_not_publish_knowledge_or_grant_authority: true,
    }),
  });
  if (!response.ok) {
    throw new ApiRequestError("Connector invocation evidence ingestion failed", response.status);
  }
  const payload: unknown = await response.json();
  const data = envelopeData(payload);
  if (!isInventoryItem(data)) throw new Error("Invocation evidence returned unsafe metadata");
  if (data.source_invocation_id !== invocation.invocation_id ||
    data.source_invocation_digest !== invocation.canonical_digest ||
    data.package_digest !== invocation.package_digest || data.capability_id !== invocation.capability_id ||
    data.capability_class !== invocation.capability_class ||
    data.required_permission !== invocation.required_permission ||
    data.normalized_redacted_result_digest !== invocation.normalized_redacted_result_digest ||
    data.ingestion_policy_id !== option.ingestion_policy_id ||
    data.ingestion_policy_digest !== option.ingestion_policy_digest ||
    data.classification !== option.classification ||
    data.retention_policy_id !== option.retention_policy_id) {
    throw new Error("Invocation evidence does not match the exact governed result");
  }
  return { data };
}

File: backend/src/atlas/modules/connectors/domain/invocation_evidence.py
from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime

from atlas.modules.identity.domain.models import AssuranceLevel, validate_stable_identifier

_DIGEST = re.compile(r"^[a-f0-9]{64}$")
ENABLED_INVOCATION_EVIDENCE_INGESTED = "enabled_invocation_evidence_ingested"


def _ids(*values: str) -> None:
    for value in values:
        validate_stable_identifier(value, "connector invocation evidence identifier")


def _digests(*values: str) -> bool:
    return all(_DIGEST.fullmatch(value) is not None for value in values)


@dataclass(frozen=True, slots=True)
class ConnectorInvocationEvidencePolicySnapshot:
    policy_id: str
    schema_version: str
    version: int
    organization_id: str
    environment_id: str
    policy_version: str
    required_source_schema: str
    required_source_state: str
    required_adapter_id: str
    required_adapter_attestor_id: str
    required_receipt_schema: str
    required_classification: str
    access_policy_id: str
    access_policy_digest: str
    retention_policy_id: str
    retention_policy_digest: str
    encryption_profile_id: str
    encryption_profile_digest: str
    maximum_source_age_minutes: int
    maximum_evidence_items: int
    maximum_evidence_bytes: int
    required_assurance_level: AssuranceLevel
    signed_by: str
    signature_verified: bool
    issued_at: datetime
    expires_at: datetime
    canonical_digest: str

    def __post_init__(self) -> None:
        _ids(
            self.policy_id,
            self.schema_version,
            self.organization_id,
            self.environment_id,
            self.policy_version,
            self.required_source_schema,
            self.required_source_state,
            self.required_adapter_id,
            self.required_adapter_attestor_id,
            self.required_receipt_schema,
            self.required_classification,
            self.access_policy_id,
            self.retention_policy_id,
            self.encryption_profile_id,
            self.signed_by,
        )
        if (
            self.version != 1
            or not 1 <= self.maximum_source_age_minutes <= 1440
            or not 1 <= self.maximum_evidence_items <= 1000
            or not 1 <= self.maximum_evidence_bytes <= 1_048_576
            or self.required_assurance_level
            not in {
                AssuranceLevel.SINGLE_FACTOR,
                AssuranceLevel.MULTI_FACTOR,
                AssuranceLevel.HARDWARE_BACKED,
            }
            or not self.signature_verified
            or self.issued_at.tzinfo is None
            or self.expires_at.tzinfo is None
            or self.expires_at <= self.issued_at
            or not _digests(
                self.access_policy_digest,
                self.retention_policy_digest,
                self.encryption_profile_digest,
                self.canonical_digest,
            )
        ):
            raise ValueError("Connector invocation evidence policy contract is invalid")


@dataclass(frozen=True, slots=True)
class ConnectorInvocationEvidenceInstruction:
    ingestion_id: str
    organization_id: str
    environment_id: str
    source_invocation_id: str
    source_invocation_digest: str
    connector_id: str
    instance_id: str
    capability_id: str
    output_schema_digest: str
    result_policy_digest: str
    normalized_redacted_result_digest: str
    source_observation_count: int
    source_output_bytes: int
    source_started_at: datetime
    source_completed_at: datetime
    classification: str
    access_policy_id: str
    access_policy_digest: str
    retention_policy_id: str
    retention_policy_digest: str
    encryption_profile_id: str
    encryption_profile_digest: str
    maximum_evidence_items: int
    maximum_evidence_bytes: int
    ingestion_policy_digest: str

    def __post_init__(self) -> None:
        _ids(
            self.ingestion_id,
            self.organization_id,
            self.environment_id,
            self.source_invocation_id,
            self.connector_id,
            self.instance_id,
            self.capability_id,
            self.classification,
            self.access_policy_id,
            self.retention_policy_id,
            self.encryption_profile_id,
        )
        if (
            not 1 <= self.source_observation_count <= self.maximum_evidence_items <= 1000
            or not 0 <= self.source_output_bytes <= self.maximum_evidence_bytes <= 1_048_576
            or self.source_started_at.tzinfo is None
            or self.source_completed_at.tzinfo is None
            or self.source_completed_at < self.source_started_at
            or not _digests(
                self.source_invocation_digest,
                self.output_schema_digest,
                self.result_policy_digest,
                self.normalized_redacted_result_digest,
                self.access_policy_digest,
                self.retention_policy_digest,
                self.encryption_profile_digest,
                self.ingestion_policy_digest,
            )
        ):
            raise ValueError("Connector invocation evidence instruction is invalid")


@dataclass(frozen=True, slots=True)
class ConnectorInvocationEvidenceReceipt:
    ingestion_id: str
    schema_version: str
    version: int
    adapter_id: str
    attested_by: str
    source_invocation_digest: str
    normalized_redacted_result_digest: str
    evidence_package_id: str
    evidence_schema_version: str
    evidence_content_digest: str
    evidence_metadata_digest: str
    classification: str
    access_policy_id: str
    access_policy_digest: str
    retention_policy_id: str
    retention_policy_digest: str
    encryption_profile_id: str
    encryption_profile_digest: str
    evidence_item_count: int
    evidence_bytes: int
    observed_from: datetime
    observed_to: datetime
    ingested_at: datetime
    immutable_storage_confirmed: bool
    encrypted_at_rest: bool
    transient_buffers_erased: bool
    artifact_channel_closed: bool
    signature_verified: bool
    canonical_digest: str

    def __post_init__(self) -> None:
        _ids(
            self.ingestion_id,
            self.schema_version,
            self.adapter_id,
            self.attested_by,
            self.evidence_package_id,
            self.evidence_schema_version,
            self.classification,
            self.access_policy_id,
            self.retention_policy_id,
            self.encryption_profile_id,
        )
        if (
            self.version != 1
            or not 1 <= self.evidence_item_count <= 1000
            or not 0 <= self.evidence_bytes <= 1_048_576
            or any(
                value.tzinfo is None
                for value in (self.observed_from, self.observed_to, self.ingested_at)
            )
            or self.observed_to < self.observed_from
            or self.ingested_at < self.observed_to
            or not all(
                (
                    self.immutable_storage_confirmed,
                    self.encrypted_at_rest,
                    self.transient_buffers_erased,
                    self.artifact_channel_closed,
                    self.signature_verified,
                )
            )
            or not _digests(
                self.source_invocation_digest,
                self.normalized_redacted_result_digest,
                self.evidence_content_digest,
                self.evidence_metadata_digest,
                self.access_policy_digest,
                self.retention_policy_digest,
                self.encryption_profile_digest,
                self.canonical_digest,
            )
        ):
            raise ValueError("Connector invocation evidence receipt is invalid")


@dataclass(frozen=True, slots=True)
class ConnectorInvocationEvidenceClaim:
    claim_id: str
    schema_version: str
    version: int
    source_invocation_id: str
    source_invocation_digest: str
    ingestion_id: str
    organization_id: str
    environment_id: str
    claimed_by: str
    purpose: str
    claimed_at: datetime
    request_binding_digest: str
    idempotency_digest: str
    canonical_digest: str

    def __post_init__(self) -> None:
        _ids(
            self.claim_id,
            self.schema_version,
            self.source_invocation_id,
            self.ingestion_id,
            self.organization_id,
            self.environment_id,
            self.claimed_by,
        )
        if (
            self.version != 1
            or not 20 <= len(self.purpose.strip()) <= 1000
            or self.claimed_at.tzinfo is None
            or not _digests(
                self.source_invocation_digest,
                self.request_binding_digest,
                self.idempotency_digest,
                self.canonical_digest,
            )
        ):
            raise ValueError("Connector invocation evidence claim is invalid")


@dataclass(frozen=True, slots=True)
class ConnectorInvocationEvidenceRecord:
    ingestion_id: str
    schema_version: str
    version: int
    claim_id: str
    source_invocation_id: str
    source_invocation_digest: str
    organization_id: str
    environment_id: str
    package_digest: str
    connector_id: str
    release_version: str
    manifest_digest: str
    instance_id: str
    instance_key: str
    display_name: str
    capability_id: str
    capability_class: str
    required_permission: str
    output_schema_digest: str
    result_policy_digest: str
    normalized_redacted_result_digest: str
    evidence_package_id: str
    evidence_schema_version: str
    evidence_content_digest: str
    evidence_metadata_digest: str
    classification: str
    access_policy_id: str
    access_policy_digest: str
    retention_policy_id: str
    retention_policy_digest: str
    encryption_profile_id: str
    encryption_profile_digest: str
    ingestion_policy_id: str
    ingestion_policy_digest: str
    ingestion_policy_version: str
    ingestion_adapter_id: str
    evidence_item_count: int
    evidence_bytes: int
    observed_from: datetime
    observed_to: datetime
    ingested_at: datetime
    instance_state: str
    ingested_by: str
    purpose: str
    canonical_digest: str
    source_invocation_completed: bool = True
    evidence_ingested: bool = True
    immutable_storage_confirmed: bool = True
    encrypted_at_rest: bool = True
    transient_buffers_erased: bool = True
    artifact_channel_closed: bool = True
    knowledge_item_created: bool = False
    retrieval_published: bool = False
    model_context_available: bool = False
    graph_updated: bool = False
    scheduled: bool = False
    workflow_continued: bool = False
    execution_authorized: bool = False
    deployment_approved: bool = False
    infrastructure_mutation_performed: bool = False
    reused: bool = False

    def __post_init__(self) -> None:
        _ids(
            self.ingestion_id,
            self.schema_version,
            self.claim_id,
            self.source_invocation_id,
            self.organization_id,
            self.environment_id,
            self.connector_id,
            self.release_version,
            self.instance_id,
            self.instance_key,
            self.capability_id,
            self.required_permission,
            self.evidence_package_id,
            self.evidence_schema_version,
            self.classification,
            self.access_policy_id,
            self.retention_policy_id,
            self.encryption_profile_id,
            self.ingestion_policy_id,
            self.ingestion_policy_version,
            self.ingestion_adapter_id,
            self.instance_state,
            self.ingested_by,
        )
        if (
            self.version != 1
            or self.instance_state != ENABLED_INVOCATION_EVIDENCE_INGESTED
            or self.capability_class not in {"C0", "C1"}
            or not 20 <= len(self.purpose.strip()) <= 1000
            or not 1 <= self.evidence_item_count <= 1000
            or not 0 <= self.evidence_bytes <= 1_048_576
            or any(
                value.tzinfo is None
                for value in (self.observed_from, self.observed_to, self.ingested_at)
            )
            or self.observed_to < self.observed_from
            or self.ingested_at < self.observed_to
            or not _digests(
                self.source_invocation_digest,
                self.package_digest,
                self.manifest_digest,
                self.output_schema_digest,
                self.result_policy_digest,
                self.normalized_redacted_result_digest,
                self.evidence_content_digest,
                self.evidence_metadata_digest,
                self.access_policy_digest,
                self.retention_policy_digest,
                self.encryption_profile_digest,
                self.ingestion_policy_digest,
                self.canonical_digest,
            )
            or not all(
                (
                    self.source_invocation_completed,
                    self.evidence_ingested,
                    self.immutable_storage_confirmed,
                    self.encrypted_at_rest,
                    self.transient_buffers_erased,
                    self.artifact_channel_closed,
                )
            )
            or any(
                (
                    self.knowledge_item_created,
                    self.retrieval_published,
                    self.model_context_available,
                    self.graph_updated,
                    self.scheduled,
                    self.workflow_continued,
                    self.execution_authorized,
                    self.deployment_approved,
                    self.infrastructure_mutation_performed,
                )
            )
        ):
            raise ValueError("Connector invocation evidence record is invalid")
